File: structy/BinaryTree/treeIncludes.js
// Write a function, treeIncludes, that takes in the root of a binary tree and a target value. 
// The function should return a boolean indicating whether or not the value is contained in the tree.

class Node {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f

function treeIncludes(root, target) {
    if (!root) return false
    if (root.val === target) return true
    return treeIncludes(root.right, target) || treeIncludes(root.left, target)
}

console.log(treeIncludes(a, "e")); // -> true

File: structy/BinaryTree/tree_includes.py
from collections import deque
class Node: 
    def __init__(self, val):
        self.val = val
        self.right = None
        self.left = None


a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')        
a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#      a
#    /   \
#   b     c
#  / \     \
# d   e     f

#! iterative breadth-first
def tree_includes(root, target):
    if not root:
        return False

    queue = deque([ root ])
    
    while queue:
        node = queue.popleft()

        if node.val == target:
            return True

        if node.left: queue.append(node.left)
        if node.right: queue.append(node.right)

    return False


#! iterative depth-first
# def tree_includes(root, target):
#     if not root:
#         return False

#     stack = [ root ]

#     while stack:
#         node = stack.pop()

#         if node.val == target:
#             return True

#         if node.right: 
#             stack.append(node.right)

#         if node.left: 
#             stack.append(node.left)
    
#     return False


#! recursive depth-first
# def tree_includes(root, target):
#     if not root:
#         return False

#     if root.val == target:
#         return True

#     return tree_includes(root.left, target) or tree_includes(root.right, target) 


print('TRUE', tree_includes(a, 'e')) 
print('FALSE', tree_includes(a, 'x')) 
